File: tests/imports/geomlib.dart
// A small geometry library imported by tests/dart-test-multifile.dart (via -i tests/imports).
// It is parsed with the Dart grammar; its class Vec and top-level function addAll register
// like the main file's own top-level declarations. There is no main() here - only the main
// file's main() is the entry point, so importing this file just registers its declarations.

class Vec {
  int x;
  int y;
  Vec(this.x, this.y);
  int dot(Vec o) {
    return this.x * o.x + this.y * o.y;
  }
  int manhattan(Vec o) {
    int dx = this.x - o.x;
    int dy = this.y - o.y;
    if (dx < 0) dx = -dx;
    if (dy < 0) dy = -dy;
    return dx + dy;
  }
  String describe() => 'Vec(${this.x}, ${this.y})';
}

int addAll(List<int> xs) {
  int total = 0;
  for (int v in xs) {
    total = total + v;
  }
  return total;
}

File: tests/imports/geomlib.ts
// A small geometry library imported by tests/typescript-test-multifile.ts (found via the
// -i include root: mec -i tests/imports ...). 'import { Vec, scale } from "./geomlib"'
// maps to this file (geomlib.ts): the leading './' is stripped and '.ts' appended. The
// file is parsed with the same grammar; its top-level class, function and const register
// like the main file's (flat, global registration). The 'export' keywords are erased at
// run time and the named-binding list on the import side is cosmetic.

export class Vec {
    x: number;
    y: number;

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }

    dot(o: Vec): number {
        return this.x * o.x + this.y * o.y;
    }

    // References Vec itself, so the imported code must be able to resolve its own class.
    add(o: Vec): Vec {
        return new Vec(this.x + o.x, this.y + o.y);
    }
}

// A top-level function that builds a Vec: exercises an imported free function.
export function scale(v: Vec, f: number): Vec {
    return new Vec(v.x * f, v.y * f);
}

// A top-level const: exercises an imported value binding.
export const ORIGIN_TAG: string = "origin";
